model/atmosphere/advection/src/icon4py/model/atmosphere/advection/advection.py

```diff
@@ -0,0 +1,443 @@
+            )
+        if not hasattr(HorizontalAdvectionLimiter, self.horizontal_advection_limiter.name):
+            raise NotImplementedError(
+                f"Horizontal advection limiter {self.horizontal_advection_limiter} not implemented."
+            )
+        if not hasattr(VerticalAdvectionType, self.vertical_advection_type.name):
+            raise NotImplementedError(
+                f"Vertical advection type {self.vertical_advection_type} not implemented."
+            )
+        if not hasattr(VerticalAdvectionLimiter, self.vertical_advection_limiter.name):
+            raise NotImplementedError(
+                f"Vertical advection limiter {self.vertical_advection_limiter} not implemented."
+            )
+
+
+class Advection(ABC):
+    """Class that runs one three-dimensional advection step."""
+
+    @abstractmethod
+    def run(
+        self,
+        diagnostic_state: advection_states.AdvectionDiagnosticState,
+        prep_adv: advection_states.AdvectionPrepAdvState,
+        p_tracer_now: fa.CellKField[ta.wpfloat],
+        p_tracer_new: fa.CellKField[ta.wpfloat],
+        dtime: ta.wpfloat,
+    ):
+        """
+        Run an advection step.
+
+        Args:
+            diagnostic_state: output argument, data class that contains diagnostic variables
+            prep_adv: input argument, data class that contains precalculated advection fields
+            p_tracer_now: input argument, field that contains current tracer mass fraction
+            p_tracer_new: output argument, field that contains new tracer mass fraction
+            dtime: input argument, the time step
+
+        """
+        ...
+
+
+class NoAdvection(Advection):
+    """Class that implements disabled three-dimensional advection."""
+
+    def __init__(
+        self,
+        grid: icon_grid.IconGrid,
+        exchange: decomposition.ExchangeRuntime = decomposition.SingleNodeExchange(),
+    ):
+        log.debug("advection class init - start")
+
+        # input arguments
+        self._grid = grid
+        self._exchange = exchange
+
+        # cell indices
+        cell_domain = h_grid.domain(dims.CellDim)
+        self._start_cell_nudging = self._grid.start_index(cell_domain(h_grid.Zone.NUDGING))
+        self._end_cell_local = self._grid.end_index(cell_domain(h_grid.Zone.LOCAL))
+
+    def run(
+        self,
+        diagnostic_state: advection_states.AdvectionDiagnosticState,
+        prep_adv: advection_states.AdvectionPrepAdvState,
+        p_tracer_now: fa.CellKField[ta.wpfloat],
+        p_tracer_new: fa.CellKField[ta.wpfloat],
+        dtime: ta.wpfloat,
+    ):
+        log.debug("advection run - start")
+
+        log.debug("communication of prep_adv cell field: mass_flx_ic - start")
+        self._exchange.exchange_and_wait(dims.CellDim, prep_adv.mass_flx_ic)
+        log.debug("communication of prep_adv cell field: mass_flx_ic - end")
+
+        log.debug("running stencil copy_cell_kdim_field - start")
```

Comment:
@halungge I would suggest we do the same this here as per https://github.com/C2SM/icon4py/pull/566

Reply:
yes, agree. @dastrm that means you pass a `backend` argument to the Advection constructors and use it in the Advection for field allocation (if applicable) and "cache" the stencils as
```
self._foo = foo.with_backend(backend)
```
using later
```
self._foo(...)
```